tests/Unit/NumericalAlgorithms/Spectral/Test_SwshDerivatives.cpp

```diff
@@ -0,0 +1,338 @@
+                               swsh_derivative_approx);
+}
+
+// This function verifies the operation of the compute_derivatives function for
+// processing collections of derivative operations which collect equal spin
+// objects during the forward and inverse transforms for transform
+// optimization. This calls the utility for two derivatives, `DerivativeKind1`
+// and `DerivativeKind2`, each applied to two scalars of spin-weight `Spin1` and
+// `Spin2`. The result is compared to analytical forms obtained by multiplying
+// the derivatives of the basis functions by the generated coefficients.
+template <ComplexRepresentation Representation, int Spin1, int Spin2,
+          typename DerivativeKind1, typename DerivativeKind2>
+void test_compute_swsh_derivatives() noexcept {
+  MAKE_GENERATOR(gen);
+  UniformCustomDistribution<size_t> size_distribution{2, 7};
+  const size_t l_max = size_distribution(gen);
+  const size_t number_of_radial_points = 2;
+  UniformCustomDistribution<double> coefficient_distribution{-10.0, 10.0};
+
+  using input_tag_list = tmpl::list<TestTag<0, Spin1>, TestTag<1, Spin2>>;
+  using derivative_tag_list =
+      tmpl::list<Tags::Derivative<TestTag<0, Spin1>, DerivativeKind1>,
+                 Tags::Derivative<TestTag<1, Spin2>, DerivativeKind1>,
+                 Tags::Derivative<TestTag<0, Spin1>, DerivativeKind2>,
+                 Tags::Derivative<TestTag<1, Spin2>, DerivativeKind2>>;
+  using collocation_variables_tag =
+      ::Tags::Variables<tmpl::append<input_tag_list, derivative_tag_list>>;
+  using coefficients_variables_tag = ::Tags::Variables<db::wrap_tags_in<
+      Tags::SwshTransform, tmpl::append<input_tag_list, derivative_tag_list>>>;
+
+  auto box = db::create<
+      db::AddSimpleTags<collocation_variables_tag, coefficients_variables_tag,
+                        Tags::LMax, Tags::NumberOfRadialPoints>,
+      db::AddComputeTags<>>(
+      db::item_type<collocation_variables_tag>{
+          number_of_radial_points * number_of_swsh_collocation_points(l_max)},
+      db::item_type<coefficients_variables_tag>{
+          size_of_libsharp_coefficient_vector(l_max) * number_of_radial_points},
+      l_max, number_of_radial_points);
+
+  ComplexModalVector expected_modes_spin_1{
+      number_of_radial_points * size_of_libsharp_coefficient_vector(l_max)};
+  TestHelpers::generate_swsh_modes<Spin1>(
+      make_not_null(&expected_modes_spin_1), make_not_null(&gen),
+      make_not_null(&coefficient_distribution), number_of_radial_points, l_max);
+
+  ComplexModalVector expected_modes_spin_2{
+      number_of_radial_points * size_of_libsharp_coefficient_vector(l_max)};
+  TestHelpers::generate_swsh_modes<Spin2>(
+      make_not_null(&expected_modes_spin_2), make_not_null(&gen),
+      make_not_null(&coefficient_distribution), number_of_radial_points, l_max);
+
+  const auto coefficients_to_analytic_collocation = [&l_max](
+      auto computed_collocation, ComplexModalVector& modes) noexcept {
```

Comment:
auto -> `const auto`

`modes` -> `expected_modes` and should become a `gsl::not_null` so it's clear at the call site this gets modified.